File: frontend/datadog_config.js
/**
 * Datadog configuration for the telecom frontend application
 */

const tracer = require('dd-trace');
const StatsD = require('hot-shots');

// Initialize Datadog tracer
function configureDatadog() {
    // Get Datadog configuration from environment variables
    const ddServiceName = process.env.DD_SERVICE || 'telecom-frontend';
    const ddEnv = process.env.DD_ENV || 'production';
    const ddVersion = process.env.DD_VERSION || '1.0.0';
    const ddAgentHost = process.env.DD_AGENT_HOST || 'datadog-agent';
    const ddTraceAgentPort = parseInt(process.env.DD_TRACE_AGENT_PORT || '8126');
    const ddDogstatsdPort = parseInt(process.env.DD_DOGSTATSD_PORT || '8125');

    // Set environment variables for dd-trace (critical for proper env detection)
    process.env.DD_SERVICE = ddServiceName;
    process.env.DD_ENV = ddEnv;
    process.env.DD_VERSION = ddVersion;
    process.env.DD_AGENT_HOST = ddAgentHost;
    process.env.DD_TRACE_AGENT_PORT = ddTraceAgentPort.toString();

    // Initialize the tracer with explicit configuration
    tracer.init({
        service: ddServiceName,
        env: ddEnv,
        version: ddVersion,
        hostname: ddAgentHost,
        port: ddTraceAgentPort,
        logInjection: true,
        analytics: true,
        analyticsSampleRate: 1.0,
        runtimeMetrics: true,
        plugins: true,
        tags: {
            service: ddServiceName,
            env: ddEnv,
            version: ddVersion,
            component: 'frontend',
            team: 'telecom',
            // Additional explicit tags to ensure environment detection
            'dd.env': ddEnv,
            'dd.service': ddServiceName,
            'dd.version': ddVersion
        }
    });

    console.log(`Datadog configured for service: ${ddServiceName}, env: ${ddEnv}, version: ${ddVersion}`);

    return tracer;
}

// Initialize StatsD client for custom metrics
function createStatsClient() {
    const ddAgentHost = process.env.DD_AGENT_HOST || 'datadog-agent';
    const ddDogstatsdPort = parseInt(process.env.DD_DOGSTATSD_PORT || '8125');

    return new StatsD({
        host: ddAgentHost,
        port: ddDogstatsdPort,
        prefix: 'telecom.frontend.',
        globalTags: {
            service: process.env.DD_SERVICE || 'telecom-frontend',
            env: process.env.DD_ENV || 'production',
            version: process.env.DD_VERSION || '1.0.0'
        }
    });
}

// Custom metrics helpers
function sendCustomMetric(statsClient, metricName, value, tags = {}) {
    try {
        statsClient.gauge(metricName, value, tags);
    } catch (error) {
        console.error(`Error sending metric ${metricName}:`, error);
    }
}

function incrementCounter(statsClient, metricName, tags = {}) {
    try {
        statsClient.increment(metricName, 1, tags);
    } catch (error) {
        console.error(`Error incrementing counter ${metricName}:`, error);
    }
}

function recordHistogram(statsClient, metricName, value, tags = {}) {
    try {
        statsClient.histogram(metricName, value, tags);
    } catch (error) {
        console.error(`Error recording histogram ${metricName}:`, error);
    }
}

function recordTiming(statsClient, metricName, duration, tags = {}) {
    try {
        statsClient.timing(metricName, duration, tags);
    } catch (error) {
        console.error(`Error recording timing ${metricName}:`, error);
    }
}

// Express middleware for request tracking
function createRequestTrackingMiddleware(statsClient) {
    return (req, res, next) => {
        const startTime = Date.now();
        
        // Track request count
        incrementCounter(statsClient, 'requests.count', {
            method: req.method,
            route: req.route ? req.route.path : req.path
        });

        // Override res.end to capture response metrics
        const originalEnd = res.end;
        res.end = function(...args) {
            const duration = Date.now() - startTime;
            
            // Track response time
            recordTiming(statsClient, 'requests.duration', duration, {
                method: req.method,
                status_code: res.statusCode,
                route: req.route ? req.route.path : req.path
            });

            // Track status codes
            incrementCounter(statsClient, 'requests.status', {
                status_code: res.statusCode,
                method: req.method
            });

            originalEnd.apply(this, args);
        };

        next();
    };
}

module.exports = {
    configureDatadog,
    createStatsClient,
    sendCustomMetric,
    incrementCounter,
    recordHistogram,
    recordTiming,
    createRequestTrackingMiddleware
};

File: backend/datadog_config.py
"""
Datadog configuration for the telecom backend application
"""
import os
from ddtrace import config, patch_all
from datadog import initialize, statsd

def configure_datadog():
    """Configure Datadog APM and metrics for the Flask application"""
    
    # Get Datadog configuration from environment variables
    dd_api_key = os.getenv('DD_API_KEY')
    dd_app_key = os.getenv('DD_APP_KEY')
    dd_service_name = os.getenv('DD_SERVICE', 'telecom-backend')
    dd_env = os.getenv('DD_ENV', 'production')
    dd_version = os.getenv('DD_VERSION', '1.0.0')
    dd_agent_host = os.getenv('DD_AGENT_HOST', 'datadog-agent')
    dd_trace_agent_port = int(os.getenv('DD_TRACE_AGENT_PORT', '8126'))
    dd_dogstatsd_port = int(os.getenv('DD_DOGSTATSD_PORT', '8125'))
    
    # Set environment variables for ddtrace (critical for proper env detection)
    os.environ['DD_SERVICE'] = dd_service_name
    os.environ['DD_ENV'] = dd_env
    os.environ['DD_VERSION'] = dd_version
    os.environ['DD_AGENT_HOST'] = dd_agent_host
    os.environ['DD_TRACE_AGENT_PORT'] = str(dd_trace_agent_port)
    
    # Configure APM
    config.flask['service_name'] = dd_service_name
    config.flask['collect_request_headers'] = True
    config.flask['collect_response_headers'] = True
    config.flask['trace_query_string'] = True
    
    # Configure service tags - CRITICAL: Set these explicitly
    config.env = dd_env
    config.service = dd_service_name
    config.version = dd_version
    
    # Set global tags to ensure environment is properly tagged
    config.tags = {
        'env': dd_env,
        'service': dd_service_name,
        'version': dd_version,
        'component': 'backend',
        'team': 'telecom'
    }
    
    # Ensure proper service mapping for APM
    config.flask['service_name'] = dd_service_name
    config.flask['distributed_tracing'] = True
    config.flask['analytics_enabled'] = True
    config.flask['analytics_sample_rate'] = 1.0
    
    # Configure trace agent
    config.trace.agent_hostname = dd_agent_host
    config.trace.agent_port = dd_trace_agent_port
    config.trace.enabled = True
    config.trace.analytics_enabled = True
    config.trace.analytics_sample_rate = 1.0
    
    # Configure specific integrations
    config.sqlalchemy['service_name'] = f"{dd_service_name}-db"
    config.requests['service_name'] = f"{dd_service_name}-http"
    
    # Patch all supported libraries
    patch_all()
    
    # Initialize Datadog metrics (optional)
    if dd_api_key and dd_app_key:
        initialize(
            api_key=dd_api_key,
            app_key=dd_app_key,
            host_name=dd_agent_host
        )
    
    # Configure StatsD for custom metrics
    statsd.host = dd_agent_host
    statsd.port = dd_dogstatsd_port
    
    print(f"Datadog configured for service: {dd_service_name}, env: {dd_env}, version: {dd_version}")

def send_custom_metric(metric_name, value, tags=None):
    """Send custom metrics to Datadog"""
    if tags is None:
        tags = []
    
    try:
        statsd.gauge(metric_name, value, tags=tags)
    except Exception as e:
        print(f"Error sending metric {metric_name}: {e}")

def increment_counter(metric_name, tags=None):
    """Increment a counter metric in Datadog"""
    if tags is None:
        tags = []
    
    try:
        statsd.increment(metric_name, tags=tags)
    except Exception as e:
        print(f"Error incrementing counter {metric_name}: {e}")

def record_histogram(metric_name, value, tags=None):
    """Record a histogram metric in Datadog"""
    if tags is None:
        tags = []
    
    try:
        statsd.histogram(metric_name, value, tags=tags)
    except Exception as e:
        print(f"Error recording histogram {metric_name}: {e}")
